clippy_lints/src/assertions_on_result_states.rs

```diff
@@ -0,0 +1,99 @@
+    /// ### Why is this bad?
+    /// An assertion failure cannot output an useful message of the error.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// # let r = Ok::<_, ()>(());
+    /// assert!(r.is_ok());
+    /// # let r = Err::<_, ()>(());
+    /// assert!(r.is_err());
+    /// ```
+    #[clippy::version = "1.64.0"]
+    pub ASSERTIONS_ON_RESULT_STATES,
+    style,
+    "`assert!(r.is_ok())`/`assert!(r.is_err())` gives worse error message than directly calling `r.unwrap()`/`r.unwrap_err()`"
+}
+
+declare_lint_pass!(AssertionsOnResultStates => [ASSERTIONS_ON_RESULT_STATES]);
+
+impl<'tcx> LateLintPass<'tcx> for AssertionsOnResultStates {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) {
+        if let Some(macro_call) = root_macro_call_first_node(cx, e)
+          && matches!(cx.tcx.get_diagnostic_name(macro_call.def_id), Some(sym::assert_macro))
+          && let Some((condition, panic_expn)) = find_assert_args(cx, e, macro_call.expn)
+          && matches!(panic_expn, PanicExpn::Empty)
+          && let ExprKind::MethodCall(method_segment, args, _) = condition.kind {
+            let method_receiver = &args[0];
+            let result_type = cx.typeck_results().expr_ty(method_receiver);
+            if method_segment.ident.name == sym!(is_ok)
+              && let Some(error_type) = get_result_child_type(cx, result_type, true /*err*/)
+              && has_debug_impl(error_type, cx)
+              && let Some(method_receiver_snippet) = snippet_opt(cx, method_receiver.span) {
+                span_lint_and_sugg(
+                    cx,
+                    ASSERTIONS_ON_RESULT_STATES,
+                    macro_call.span,
+                    "called `assert!` with `Result::is_ok`",
+                    "replace with",
+                    format!(
+                        "{}.unwrap()",
+                        method_receiver_snippet
+                    ),
+                    Applicability::MaybeIncorrect,
+                );
+            } else if method_segment.ident.name == sym!(is_err)
+              && let Some(value_type) = get_result_child_type(cx, result_type, false /*err*/)
+              && has_debug_impl(value_type, cx)
+              && let Some(method_receiver_snippet) = snippet_opt(cx, method_receiver.span) {
+                span_lint_and_sugg(
+                    cx,
+                    ASSERTIONS_ON_RESULT_STATES,
+                    macro_call.span,
+                    "called `assert!` with `Result::is_err`",
+                    "replace with",
+                    format!(
+                        "{}.unwrap_err()",
+                        method_receiver_snippet
+                    ),
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        }
+    }
+}
+
+/// Given a `Result<T, E>` type, return its value type (`T`) or error type (`E`).
+fn get_result_child_type<'a>(cx: &LateContext<'_>, ty: Ty<'a>, err: bool) -> Option<Ty<'a>> {
```

Comment:
In general, functions taking boolean parameters to change there behavior are best avoided. For short functions it can be better to split them into two functions unless you need the ability to dispatch at runtime.

For this case it would be better to inline the function. e.g.
```rust
if ...
    && if let ty::Adt(def, substs) = ty.kind()
    && cx.tcx.is_diagnostic_item(sym::Result, def.did())
{
    if name == "is_ok"
        && let Some(ty) = substs.type_at(0)
        ...
    {}
}
```